### ISSUE TITLE
Add an option for torchrun c10d store to be collocated with rank 0 training node

### ISSUE THREAD

**Issue by chajath** [pytorch/pytorch#118891]
### 🚀 The feature, motivation and pitch

When running elastic distributed training with `torchrun` and `c10d` rendezvous backend, node ranks are designated by `c10d` store backend and are usually different node to the `c10d` store leader node.

I have a job where rank 0 node takes substantially more time to finish on train end hook, as closing fd handler takes time when using in conjunction with tools like `gcsfuse`. I expect training nodes with rank>0 to exit early, and it is a good thing as it frees up GPU nodes for other uses.

However, this causes `c10d` store node to exit early and tear down rendezvous server after `exit_barrier_timeout` (which is hardwired to be 300s). This causes cascading failure to the long-running rank 0 node, potentially causing data loss at the end of the training job.

One possible remedy could be to have an option so that the non-standalone `torchrun` with designated `c10d` store always assumes rank 0. This way, the lifeline of `c10d` store server will coincide with the long-running rank 0, preventing cascading failure. 

### Alternatives

We could make `exit_barrier_timeout` configurable. However, this is a brittle logic and is probably not the intended use of `exit_barrier_timeout`.

Using `static` backend can work. However, we don't benefit as much fail-over elasticity as `c10d` backend

**Comment by wconstab** [comment]
cc @kurman

**Comment by kurman** [comment]
Have you tried using etcd based rendezvous? Having an external store would remove that dependency and keep the elasticity. The downside is an additional dependency.

Going further into design, this falls into topology discussion and we need to have an ability to control rank assignment for multiple reasons and is one of the usecases and we are planning to work on that.

**Comment by chajath** [comment]
Hi @kurman, I've tried `etcd` backend and it worked. But as you have mentioned, I'm still wary of the fact that it's adding additional dependency. On `k8s` it is easy enough, but it gets a little more laborious on more minimalistic setup.

I think the whole topology concept will be directly useful to my use case where deployment topology actually matters a lot. However I'm also wondering if there's an appetite to treat this issue as a one-off fix we can work on, and can potentially put behind a `c10d` storage backend specific flag. It if's not too complicated I would be happy to give it a go.
